headingoffset-polyfill.js

```diff
@@ -1,157 +1,268 @@
 // @ts-check
 
+// Track headings with polyfill-managed 'aria-level' attributes
 const managedHeadings = new WeakSet();
 
-const ariaLevelObserver = new MutationObserver(function (mutationList) {
-  for (const mutation of mutationList) {
-    if (
-      mutation.target instanceof HTMLHeadingElement
-    ) {
-      // 'aria-level' was modified
-      // 'aria-level' was added
-      if (mutation.target.hasAttribute("aria-level")) {
-        stopManagingHeading(mutation.target); // (it’ll get re-added if the polyfill made this change)
-      }
-      // 'aria-level' was removed
-      else {
-        startManagingHeading(mutation.target);
-        applyHeadingOffset(mutation.target);
-      }
+// Track observed roots
+const observedRoots = new WeakSet();
+
+/** Handle 'aria-level' attribute changes */
+const ariaLevelObserver = new MutationObserver((mutations) => {
+  for (const mutation of mutations) {
+    if (!(mutation.target instanceof HTMLHeadingElement)) continue;
```

Comment:
It might provide more clarity here to check the mutation type. It's not necessary but it'd make the code more obvious:

```suggestion
    if (!(mutation.target instanceof HTMLHeadingElement)) continue;
    if (mutation.type !== "attributes") continue;
```